File: src/okbay/desks.py
"""Standing desks: curate / work / code / deck. Tiny router, no bandit."""
from __future__ import annotations
import json, time
from pathlib import Path
from . import paths

DESK_KINDS = ("curate", "work", "code", "deck")

def _load(ws: Path | None = None) -> dict:
    p = paths.desks_path(ws)
    if p.exists():
        try:
            return json.loads(p.read_text(encoding="utf-8"))
        except json.JSONDecodeError:
            pass
    return {"active": None, "history": []}

def _save(data: dict, ws: Path | None = None) -> None:
    p = paths.desks_path(ws)
    p.parent.mkdir(parents=True, exist_ok=True)
    p.write_text(json.dumps(data, indent=2), encoding="utf-8")

def start(kind: str, objective: str = "", ws: Path | None = None) -> dict:
    kind = kind.lower().strip()
    if kind not in DESK_KINDS:
        raise ValueError(f"unknown desk {kind}; want {DESK_KINDS}")
    data = _load(ws)
    desk = {
        "id": kind,
        "state": "working",
        "objective": objective,
        "seated_at": time.time(),
        "fanout": 2 if kind in {"work", "code"} else 1,
        "layout": {"curate": "n=1", "work": "hsl 2", "code": "hdl", "deck": "web-app"}.get(kind),
    }
    data["active"] = desk
    data["history"].append({"id": kind, "objective": objective, "at": desk["seated_at"]})
    _save(data, ws)
    bb = {"claims": [], "desk": kind, "objective": objective}
    paths.blackboard_path(ws).parent.mkdir(parents=True, exist_ok=True)
    paths.blackboard_path(ws).write_text(json.dumps(bb, indent=2), encoding="utf-8")
    return desk

def stop(kind=None, ws: Path | None = None) -> dict:
    data = _load(ws)
    if data.get("active"):
        data["active"]["state"] = "quiet"
    _save(data, ws)
    return data.get("active") or {}

def dismiss(ws: Path | None = None) -> None:
    data = _load(ws)
    data["active"] = None
    _save(data, ws)

def status(ws: Path | None = None) -> dict | None:
    return _load(ws).get("active")

def blackboard(ws: Path | None = None) -> dict:
    p = paths.blackboard_path(ws)
    if p.exists():
        try:
            return json.loads(p.read_text(encoding="utf-8"))
        except json.JSONDecodeError:
            pass
    return {"claims": []}

def add_claim(claim: str, evidence: list[str] | None = None, provenance: str = "", ws: Path | None = None) -> dict:
    bb = blackboard(ws)
    rec = {"claim": claim, "evidence": evidence or [], "provenance": provenance, "verdict": "unclassified", "ts": time.time()}
    bb.setdefault("claims", []).append(rec)
    paths.blackboard_path(ws).write_text(json.dumps(bb, indent=2), encoding="utf-8")
    return rec

def blackboard_read(ws=None):
    return blackboard(ws)

def blackboard_write(claim, evidence=None, provenance="", ws=None):
    return add_claim(claim, evidence=evidence, provenance=str(provenance), ws=ws)

File: crates/okbayd/src/desks.rs
use crate::paths;
use serde_json::{json, Value};
use std::fs;
use std::path::Path;
use std::time::{SystemTime, UNIX_EPOCH};

const KINDS: &[&str] = &["curate", "work", "code", "deck"];

fn load(ws: &Path) -> Value {
    let p = paths::desks_path(ws);
    if let Ok(text) = fs::read_to_string(p) {
        if let Ok(v) = serde_json::from_str::<Value>(&text) { return v; }
    }
    json!({"active": null, "history": []})
}
fn save(ws: &Path, data: &Value) {
    let p = paths::desks_path(ws);
    if let Some(parent) = p.parent() { let _ = fs::create_dir_all(parent); }
    let _ = fs::write(p, serde_json::to_string_pretty(data).unwrap_or_else(|_| "{}".into()));
}
pub fn start(ws: &Path, kind: &str, objective: &str) -> Value {
    let kind = kind.trim().to_lowercase();
    if !KINDS.contains(&kind.as_str()) {
        return json!({"ok": false, "error": format!("unknown desk {kind}")});
    }
    let layout = match kind.as_str() {
        "curate" => "n=1", "work" => "hsl 2", "code" => "hdl", "deck" => "web-app", _ => "",
    };
    let seated = SystemTime::now().duration_since(UNIX_EPOCH).map(|d| d.as_secs_f64()).unwrap_or(0.0);
    let desk = json!({"id": kind, "state": "working", "objective": objective, "seated_at": seated, "layout": layout});
    let mut data = load(ws);
    data["active"] = desk.clone();
    save(ws, &data);
    desk
}
pub fn stop(ws: &Path) -> Value {
    let mut data = load(ws);
    if !data["active"].is_null() { data["active"]["state"] = json!("quiet"); }
    save(ws, &data);
    data["active"].clone()
}
pub fn status(ws: &Path) -> Value {
    load(ws).get("active").cloned().unwrap_or(json!(null))
}
